Add flexibility to argument check in docker entrypoint

motdotla opened the pull request:
This allows prepending commands to `./bin/rails server` as the Docker CMD when running this container, e.g. `some-binary-or-script -- ./bin/rails server -b 0.0.0.0`.

### Motivation / Background

This Pull Request has been created because I wanted to prepend the `bin/rails server` command with another binary `dotenvx run --` but have it still run `db:prepare` as expected.

```
# Dockerfile
...
CMD ["dotenvx", "run", "--debug", "--", "./bin/rails", "server"]
```

### Detail

This Pull Request changes the docker entrypoint, so `bin/rails server` (or `bin/rails s`) can be found at any location in the run `CMD` for checking if we should run the db:prepare step.

### Additional information

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* N/A Tests are added or updated if you fix a bug or add a feature.
* N/A CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

motdotla committed:
Add flexibility to argument check in docker entrypoint

zzak commented:
Thanks for your PR!

Two quick things that crossed my mind:

The `bin/rails s` short-cut seems unnecessary here as the command is something you usually only write once as part of your other deployment configuration, if at all.

If we want to allow pre-pending commands to the container startup, this should be a first-class feature with tests to cover the various use-cases and there has to be a good enough reason to maintain this complexity inside Rails. I think in this case, you have a identified a problem doesn't impact the majority of applications and can be resolved by modifying the startup as you see fit, that is the purpose of these generated files -- just to get you on the ground running, not to solve every possible deployment scenario.

That said though, if something like [kamal](https://kamal-deploy.org/docs/configuration) depends on `dotenv` and this is a blocker, perhaps it should be investigated to determine whether this is something we need to change on Rails end.

motdotla commented:
> doesn't impact the majority of applications and can be resolved by modifying the startup

Yes, good point.

What do you think about me adding an `else` statement to this template - echoing out a message like `skipping db:test:prepare` and exit early.

Here's my reasoning:

It would self-educate me and others that modifying `CMD` could affect your data migrations.

My scenario was this:

1. Deployed with kamal to hetzner - all good
2. Modified my Dockerfile `CMD` with the dotenvx prepended bit and deployed - all good
3. Added a new table and deployed - !missing table. wasn't sure why.

It took me embarrassingly an hour to track down my problem. That's when I finally looked in `docker-entrypoint.sh` and facepalmed 🤦 , leading to this PR.

zzak commented:
I can understand your frustration, thank you for explaining the situation!

Looking at #46778, it seems the inspiration for this is to allow running other commands using the same image, so printing a warning for commands that don't need `db:prepare` maybe distracting.

Now I'm curious why `db:prepare` is really there, to make `rails new && docker run` as simple as possible?
You can click "db:migrate" in the development error page, I'm not sure.
